presto-main/src/main/java/com/facebook/presto/execution/scheduler/group/DynamicLifespanScheduler.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.scheduler.group;
+
+import com.facebook.presto.execution.Lifespan;
+import com.facebook.presto.execution.scheduler.BucketNodeMap;
+import com.facebook.presto.execution.scheduler.SourceScheduler;
+import com.facebook.presto.spi.Node;
+import com.facebook.presto.spi.connector.ConnectorPartitionHandle;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.SettableFuture;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntListIterator;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.OptionalInt;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class DynamicLifespanScheduler
+        implements LifespanScheduler
+{
+    private final BucketNodeMap bucketNodeMap;
+    private final List<Node> allNodes;
+    private final List<ConnectorPartitionHandle> partitionHandles;
+    private final OptionalInt concurrentLifespansPerTask;
+
+    private final IntListIterator driverGroups;
+
+    private boolean initialScheduled;
+    private SettableFuture<?> newDriverGroupReady = SettableFuture.create();
+    @GuardedBy("this")
+    private final List<Lifespan> recentlyCompletedDriverGroups = new ArrayList<>();
+
+    public DynamicLifespanScheduler(BucketNodeMap bucketNodeMap, List<Node> allNodes, List<ConnectorPartitionHandle> partitionHandles, OptionalInt concurrentLifespansPerTask)
+    {
+        this.bucketNodeMap = bucketNodeMap;
+        this.allNodes = allNodes;
+        this.partitionHandles = requireNonNull(partitionHandles, "partitionHandles is null");
+
+        if (concurrentLifespansPerTask.isPresent()) {
```

Comment:
I think you can use `concurrentLifespansPerTask.ifPresent(lifespansPerTask ->checkArgument( ...)`